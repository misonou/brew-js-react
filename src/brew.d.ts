declare namespace Brew {
    interface PageEvent {
        /**
         * Returns the view component associated to target.
         * The property is only available in `pageenter` or `pageleave event.
         */
        readonly view?: import("./view").ViewComponent<any>;
    }
}
