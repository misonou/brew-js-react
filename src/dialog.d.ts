export type DialogCloseCallback<T> = (value?: T) => void;

export interface DialogState<T> {
    readonly root: HTMLElement;
    open: () => Promise<T | undefined>;
    close: DialogCloseCallback<T>;
}

export interface DialogBaseProps<T> {
    title?: string;
    className?: string;
    showCloseButton?: boolean;
    preventLeave?: boolean;
    modal?: boolean;
    onCommit?: (value: T | undefined) => void;
    onOpen?: (root: HTMLElement) => void;
    onClose?: (root: HTMLElement) => void;
}

export interface DialogRenderComponentProps<T> extends DialogBaseProps<T> {
    closeDialog: DialogCloseCallback<T>;
}

export interface DialogOptions<T> extends DialogBaseProps<T> {
    onRender: React.FC<DialogRenderComponentProps<T>>;
}

export interface DialogProps<T> extends React.PropsWithChildren<DialogBaseProps<T>> {
    isOpen: boolean;
}

export function createDialog<T = any>(props: DialogOptions<T>): DialogState<T>;

export function Dialog<T>(props: DialogProps<T>): JSX.Element;
