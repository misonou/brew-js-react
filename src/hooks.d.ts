/**
 * Returns if the app has completed initialization, and
 * refresh the component when the app has changed its ready state.
 * @see {@link Brew.AppInstance.ready}
 */
export function useAppReady(): boolean;

/**
 * Returns the current language, and
 * refresh the component when the language has changed.
 */
export function useLanguage(): string;

/**
 * Returns the value of specified route parameter, and
 * refresh the component when the route parameter has changed.
 * @param name Name of route parameter.
 * @param defaultValue When given, an empty value in current route state will cause a redirection which set the route parameter to such value.
 */
export function useRouteParam(name: string, defaultValue?: string): string;
