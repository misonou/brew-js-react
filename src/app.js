import { combineFn, throwNotFunction } from "zeta-dom/util";
import { addExtension, install } from "brew-js/app";
import defaults from "brew-js/defaults";

/** @type {Brew.AppInstance<Brew.WithRouter & Brew.WithI18n>} */
export var app;

const appInitCallabcks = [];

export function onAppInit(callback) {
    if (app) {
        callback(app);
    } else {
        appInitCallabcks.push(throwNotFunction(callback));
    }
}

const extension = addExtension(true, 'react2', ['?router', '?i18n'], function (app_) {
    app = app_;
    combineFn(appInitCallabcks)(app);
});

install('react', extension);
defaults.react = true;
