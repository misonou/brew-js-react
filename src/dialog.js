import { createElement, useEffect, useState } from "react";
import ReactDOM from "react-dom";
import ReactDOMClient from "./include/external/react-dom-client.js";
import { always, either, extend, makeAsync, noop, pipe, setImmediate } from "./include/zeta-dom/util.js";
import { containsOrEquals, removeNode, setClass } from "./include/zeta-dom/domUtil.js";
import dom from "./include/zeta-dom/dom.js";
import { lock, notifyAsync, preventLeave, subscribeAsync } from "./include/zeta-dom/domLock.js";
import { closeFlyout, openFlyout } from "./include/brew-js/domAction.js";

/**
 * @param {Partial<import("./dialog").DialogOptions<any>>} props
 */
export function createDialog(props) {
    var root = document.createElement('div');
    var reactRoot = ReactDOMClient.createRoot(root);
    var closeDialog = closeFlyout.bind(0, root);
    var promise;

    dom.on(root, {
        flyouthide: function () {
            removeNode(root);
            (props.onClose || noop)(root);
            if (props.onRender) {
                reactRoot.unmount();
            }
        },
        asyncStart: function () {
            setClass(root, 'loading', true);
        },
        asyncEnd: function () {
            setClass(root, 'loading', false);
        }
    });
    subscribeAsync(root);

    return {
        root: root,
        close: closeDialog,
        open: function () {
            if (promise) {
                return promise;
            }
            root.className = props.className || '';
            document.body.appendChild(root);
            dom.retainFocus(dom.activeElement, root);
            if (props.modal) {
                root.setAttribute('is-modal', '');
                dom.setModal(root);
            }
            if (props.onRender) {
                var dialogProps = extend({}, props, {
                    closeDialog: function (value) {
                        var promise = makeAsync(props.onCommit || pipe)(value);
                        notifyAsync(dom.activeElement, promise);
                        return promise.then(closeDialog);
                    }
                });
                var content = createElement(props.onRender, dialogProps);
                if (props.wrapper) {
                    content = createElement(props.wrapper, dialogProps, content);
                }
                reactRoot.render(content);
                setImmediate(function () {
                    dom.focus(root);
                });
            }
            promise = openFlyout(root);
            if (props.preventLeave) {
                preventLeave(root, promise);
            } else if (props.preventNavigation) {
                lock(root, promise);
            }
            always(promise, function () {
                promise = null;
            });
            (props.onOpen || noop)(root);
            return promise;
        }
    };
}

/**
 * @param {import("./dialog").DialogProps} props
 */
export function Dialog(props) {
    const _props = useState({})[0];
    const dialog = useState(function () {
        return createDialog(_props);
    })[0];
    extend(_props, props);

    useEffect(function () {
        var opened = containsOrEquals(dom.root, dialog.root);
        if (either(opened, _props.isOpen)) {
            if (!opened) {
                dialog.open();
            } else {
                dialog.close();
            }
        }
    }, [_props.isOpen])
    useEffect(function () {
        return dialog.close;
    }, [dialog]);
    return ReactDOM.createPortal(props.children, dialog.root);
}
