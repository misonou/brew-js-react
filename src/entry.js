import { defineGetterProperty } from "zeta-dom/util";

defineGetterProperty(window, 'brew-js-react', function () {
    console.warn('window["brew-js-react"] is deprecated, access brew.react instead.');
    return brew.react;
});

export * from "./index.js";
