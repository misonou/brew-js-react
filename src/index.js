import brew from "./include/brew-js/app.js";
export default brew;

export * from "./dialog.js";
export * from "./hooks.js";
export * from "./i18n.js";
export * from "./mixin.js";
export * from "./view.js";
