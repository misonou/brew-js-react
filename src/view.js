import React from "react";
import { useAsync } from "zeta-dom-react";
import { any, definePrototype, each, exclude, extend, isFunction, keys, makeArray, noop, pick, randomId, setImmediate } from "./include/zeta-dom/util.js";
import { animateIn, animateOut } from "./include/brew-js/anim.js";
import { app } from "./app.js";

const routeMap = new Map();
const usedParams = {};

let stateId;

function ViewContainer() {
    /** @type {any} */
    var self = this;
    React.Component.apply(self, arguments);
    self.mounted = false;
    if (!stateId) {
        stateId = history.state;
        app.on('navigate', function () {
            stateId = history.state;
        });
    }
    self.componentWillUnmount = app.on('navigate', function () {
        if (self.mounted && self.getViewComponent()) {
            self.forceUpdate();
        }
    });
}

definePrototype(ViewContainer, React.Component, {
    componentDidMount: function () {
        this.mounted = true;
    },
    render: function () {
        /** @type {any} */
        var self = this;
        var V = self.getViewComponent();
        if (V && V !== self.currentViewComponent) {
            self.currentViewComponent = V;
            if (self.currentView && self.currentElement) {
                self.prevView = self.currentView;
                self.prevElement = self.currentElement;
                self.currentElement = undefined;
                animateOut(self.prevElement, 'show').then(function () {
                    self.prevElement = undefined;
                    self.prevView = undefined;
                    self.forceUpdate();
                });
            }
            self.currentView = React.createElement(V, {
                key: routeMap.get(V).id,
                rootProps: self.props.rootProps,
                onComponentLoaded: function (element) {
                    self.currentElement = element;
                    setImmediate(function () {
                        return animateIn(element, 'show');
                    });
                }
            });
        }
        return React.createElement(React.Fragment, null, self.prevView, self.currentView);
    },
    getViewComponent: function () {
        var props = this.props;
        return any(props.views, isViewMatched) || (history.state === stateId && void redirectTo(props.defaultView));
    }
});

export function isViewMatched(view) {
    var params = routeMap.get(view);
    return !!params && false === any(params.matchers, function (v, i) {
        var value = app.route[i] || '';
        return isFunction(v) ? !v(value) : (v || '') !== value;
    });
}

export function registerView(factory, routeParams) {
    var Component = function (props) {
        var Component = useAsync(factory)[0];
        return React.createElement('div', extend({}, props.rootProps, {
            ref: function (element) {
                if (element && Component) {
                    (props.onComponentLoaded || noop)(element);
                }
            },
            children: Component && React.createElement(Component.default)
        }));
    };
    routeParams = extend({}, routeParams);
    each(routeParams, function (i, v) {
        usedParams[i] = true;
        if (v instanceof RegExp) {
            routeParams[i] = v.test.bind(v);
        }
    });
    routeMap.set(Component, {
        id: randomId(),
        matchCount: keys(routeParams).length,
        matchers: routeParams,
        params: pick(routeParams, function (v) {
            return typeof v === 'string';
        })
    });
    return Component;
}

export function renderView() {
    var views = makeArray(arguments);
    var rootProps = isFunction(views[0]) ? {} : views.shift();
    var defaultView = views[0];
    views.sort(function (a, b) {
        return (routeMap.get(b) || {}).matchCount - (routeMap.get(a) || {}).matchCount;
    });
    return React.createElement(ViewContainer, { rootProps, views, defaultView });
}

export function linkTo(view, params) {
    var viewParams = (routeMap.get(view) || {}).params;
    var newParams = {};
    for (var i in app.route) {
        if (viewParams && i in viewParams) {
            newParams[i] = viewParams[i];
        } else if (params && i in params) {
            newParams[i] = params[i];
        } else if (!usedParams[i]) {
            newParams[i] = app.route[i];
        }
    }
    return app.route.getPath(newParams);
}

export function navigateTo(view, params) {
    return app.navigate(linkTo(view, params));
}

export function redirectTo(view, params) {
    return app.navigate(linkTo(view, params), true);
}
