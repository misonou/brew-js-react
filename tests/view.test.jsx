import React from "react";
import { render, screen } from "@testing-library/react";
import { linkTo, registerView, renderView } from "src/view";
import { initApp } from "./testUtil";

/** @type {Brew.AppInstance<Brew.WithRouter>} */
let app;
/** @type {import("src/view").ViewComponent<{}>} */
let Foo;
/** @type {import("src/view").ViewComponent<{}>} */
let Bar;

beforeAll(async () => {
    app = await initApp(app => {
        app.useRouter({
            baseUrl: '/',
            routes: [
                '/{view:foo}/{baz?}',
                '/{view}/*',
                '/*'
            ]
        });
    });

    Foo = registerView(async () => {
        return {
            default: () => {
                return (<div>foo</div>)
            }
        }
    }, { view: 'foo' })

    Bar = registerView(async () => {
        return {
            default: () => {
                return (<div>bar</div>)
            }
        }
    }, { view: 'bar' })
});

beforeEach(async () => {
    await app.navigate(app.initialPath);
});

describe('renderView', () => {
    it('should render matched view', async () => {
        await app.navigate('/bar');
        const { asFragment } = render(<div>{renderView(Foo, Bar)}</div>)
        await screen.findByText('bar');
        expect(asFragment()).toMatchSnapshot();
    });

    it('should cause redirection to the first match view if none was matched', async () => {
        const { asFragment } = render(<div>{renderView(Foo, Bar)}</div>)
        await screen.findByText('foo');
        expect(asFragment()).toMatchSnapshot();
        expect(app.path).toBe('/foo');
    });

    it('should pass props to view container', async () => {
        const { asFragment } = render(<div>{renderView({ className: 'foo' }, Foo)}</div>)
        await screen.findByText('foo');
        expect(asFragment()).toMatchSnapshot();
    });
});

describe('linkTo', () => {
    it('should return path that will render specified view with params', () => {
        expect(linkTo(Foo, { baz: 'baz' })).toBe('/foo/baz');
    });
});
