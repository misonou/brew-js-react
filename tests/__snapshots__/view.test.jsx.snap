// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renderView should cause redirection to the first match view if none was matched 1`] = `
<DocumentFragment>
  <div>
    <div>
      <div>
        foo
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`renderView should pass props to view container 1`] = `
<DocumentFragment>
  <div>
    <div
      class="foo"
    >
      <div>
        foo
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`renderView should render matched view 1`] = `
<DocumentFragment>
  <div>
    <div>
      <div>
        bar
      </div>
    </div>
  </div>
</DocumentFragment>
`;
